### PR TITLE
Support Batch Reading of Messages

### PR DESCRIPTION
**Pull request by lukemerrett**
Allow us to configure sending a batch of messages to a decorated call handler.

The workers already pull messages in batches by default (10 messages max over a 10 second long poll), this change allows us to send all of those messages to a function to be processed in one go.

Intended for scenarios where batching up messages (say ingesting data into a database, or search index) would be more performant than doing it individually

Example usage:

```python
sqs = SQSEnv()

create_standard_queue(sqs, "emails")

queue = sqs.queue("emails", batching_policy=BatchMessages(batch_size=10))

@queue.processor("send_email")
def send_email(messages: list):
    for email in messages:
        print(f"Sending email {email['subject']} to {email['to']}")
```

**Important Notes:**

* Configured at **queue** level, not processor level
    * This is as we already pick up messages in batch from a queue then distribute them across all configured jobs
* If the call handler fails, all messages in the batch are discarded and requeued based on the configured queue visibility and retry policy
    * As such the call handler should expect duplicates and be idemponent
* Changes the internals of `Processor`
    * As we encourage inheritance and customisation of the `Processor` in our README, is it worth making this a major version bump to denote breaking changes (do we use semantic versioning in this project?)

### CODE REVIEW

File: sqs_workers/processors.py [hunk 1]
```diff
@@ -59,11 +59,49 @@ def process_message(self, message):
...
+        """
+        message_ids = [m.message_id for m in messages]
+        extra = {
+            "message_ids": message_ids,
+            "queue_name": self.queue.name,
+            "job_name": self.job_name,
+        }
+        logger.debug(
+            "Processing batch for {queue_name}.{job_name}".format(**extra), extra=extra
+        )
```
[the remark this answers is not in the document]
Comment (lukemerrett) [reply sqs_workers/processors.py:75]: [This the the level the other log messages are outputting as in `sqs_workers`.
](https://github.com/Doist/sqs-workers/blob/main/sqs_workers/processors.py#L44)
I could make it configurable?  Retain the default `logger.debug` for existing usage, but add a flag letting us output as `logger.info`?

File: CHANGELOG [hunk 2]
```diff
@@ -1,3 +1,8 @@
+## [0.6.0] - 2022-08-25
+
+- Added ability to configure a queue to send batches of messages
+  to a handling function rather than processing them individually
+
```
Comment (Schnouki) [inline CHANGELOG:5]: Please use the [Keep a changelog](https://keepachangelog.com/en/1.0.0/) convention:
```suggestion
## [Unreleased]

### Added

- Added ability to configure a queue to send batches of messages
  to a handling function rather than processing them individually

```

File: sqs_workers/__version__.py [hunk 3]
```diff
@@ -1,4 +1,4 @@
-VERSION = (0, 5, 6)
+VERSION = (0, 6, 0)
```
Comment (Schnouki) [inline sqs_workers/__version__.py:1]: Nope, that should be in a separate PR :)

File: sqs_workers/processors.py (hunk 1 above)
Comment (Schnouki) [inline sqs_workers/processors.py:75]: It's bad practice to format log strings: keeping "%s" etc. in them allows logging systems like Sentry or DataDog to group similar messages.
```suggestion
        logger.debug(
            "Processing batch for %s.%s",
            self.queue_name,
            self.job_name,
             extra=extra,
        )
```

File: sqs_workers/processors.py [hunk 4]
```diff
@@ -59,11 +59,49 @@ def process_message(self, message):
...
+
+        try:
+            self.process_batch(messages)
+        except Exception:
+            logger.exception(
+                "Error while processing batch for {queue_name}.{job_name}".format(
```
Comment (Schnouki) [inline sqs_workers/processors.py:81]: Same.

File: sqs_workers/queue.py [hunk 5]
```diff
@@ -242,10 +281,41 @@ def process_message(self, message):
...
+        """
+        Sends a list of messages to the call handler
+
+        Return True if processing went successful for all messages in the batch.
+        """
+        message_ids = [m.message_id for m in messages]
+        extra = {"message_ids": message_ids, "queue_name": self.name}
+        if not self.processor:
+            logger.warning(
+                "No processor set for {queue_name}".format(**extra), extra=extra
```
Comment (Schnouki) [inline sqs_workers/queue.py:295]: Same ×3 :)

File: sqs_workers/sqs_env.py [hunk 6]
```diff
@@ -50,35 +50,43 @@ def queue(
...
+        queue_name: str,
+        job_name: str,
+        pass_context=False,
+        context_var=DEFAULT_CONTEXT_VAR,
     ):
         """
         Decorator to attach processor to all jobs "job_name" of the queue "queue_name".
         """
-        q = self.queue(queue_name, queue_maker=JobQueue)  # type: JobQueue
+        q: JobQueue = self.queue(queue_name, queue_maker=JobQueue)  # type: ignore
```
Comment (Schnouki) [inline sqs_workers/sqs_env.py:80]: You can remove `# type: ...` comments if you're using type annotations. We're not trying to support Python 2.7 anymore anyway 🙂

File: sqs_workers/async_task.py [hunk 7]
```diff
@@ -30,8 +30,11 @@ def run(self, *args, **kwargs):
         """
         Run the task synchronously.
         """
-        kwargs = bind_arguments(self.processor, args, kwargs)
-        return self.processor(**kwargs)
+        if self.queue.batching_policy.batching_enabled:
+            return self.processor([kwargs])
+        else:
+            kwargs = bind_arguments(self.processor, args, kwargs)
+            return self.processor(**kwargs)
```
Comment (Schnouki) [inline sqs_workers/async_task.py:37]: Something seems wrong here 🤔

- Without batching: we use `bind_arguments()` to convert `(*args, **kwargs)` to a single `kwargs` dict, which we can then pass to the task function as `function(**kwargs)`
- With batching: we pass raw kwargs... and just ignore positional arguments!?

Maybe that's what we want. But in that case, I'd prefer to fail loudly if `args` is not empty, so that developers immediately realize that.

File: sqs_workers/async_task.py [hunk 8]
```diff
@@ -49,7 +52,8 @@ def delay(self, *args, **kwargs):
         _delay_seconds = kwargs.pop("_delay_seconds", None)
         _deduplication_id = kwargs.pop("_deduplication_id", None)
         _group_id = kwargs.pop("_group_id", None)
-        kwargs = bind_arguments(self.processor, args, kwargs)
+        if not self.queue.batching_policy.batching_enabled:
+            kwargs = bind_arguments(self.processor, args, kwargs)
```
Comment (Schnouki) [inline sqs_workers/async_task.py:56]: Same here. If we ignore `*args` when batching is enabled, please make it super explicit.

File: sqs_workers/async_task.py (hunk 7 above)
Comment (Schnouki) [reply sqs_workers/async_task.py:37]: Maybe we should even use something like `kwargs = bind_arguments(self.processor, [[kwargs]], {})` if batching is enabled, to also fail loudly if `self.processor` doesn't accept a single positional argument.

**Review by Schnouki — CHANGES_REQUESTED** [review]
Looks mostly good to me. Thanks for this, Luke! I'm just a bit concerned about how `*args` is silently ignored when batching is enabled; the rest is far less important.

**Comment by lukemerrett** [comment]
Thank you @Schnouki & @amleczko, processing the feedback now

File: CHANGELOG (hunk 2 above)
Comment (lukemerrett) [reply CHANGELOG]: Corrected, also added a `CONTRIBUTING.md` and copied over the parts from the readme along with notes on the changelog format so I won't forget next time

File: sqs_workers/__version__.py (hunk 3 above)
Comment (lukemerrett) [reply sqs_workers/__version__.py]: 👍 Done

File: sqs_workers/processors.py (hunk 1 above)
Comment (amleczko) [inline sqs_workers/processors.py:75]: Why not go even simplier:

```suggestion
        logger.debug("Processing batch", extra=extra)
```

Comment (lukemerrett) [reply sqs_workers/processors.py]: Done, this style of logging was scattered across the project so adjusted them all now to use this type of interpolation.

File: sqs_workers/sqs_env.py (hunk 6 above)
Comment (lukemerrett) [reply sqs_workers/sqs_env.py:80]: Thank goodness for that, I didn't want to rock the boat, but not a fan 😊

`# type: ignore` is intentional here, getting `mypy` warnings as it can't work out that passing `queue_maker=JobQueue` into `self.queue` always makes it return a `JobQueue` type, and the `processor` function called afterwards doesn't exist on the base class (nor should it as `RawQueue` has a different signature)

Comment (lukemerrett) [reply sqs_workers/sqs_env.py:79]: Done, moved all typing to annotations over comments

File: sqs_workers/processors.py (hunk 1 above)
Comment (lukemerrett) [reply sqs_workers/processors.py]: Would clean up the logging statements for sure, I'm keen for this, happy to go back and change again.

@Schnouki thoughts?

File: sqs_workers/async_task.py (hunk 7 above)
Comment (lukemerrett) [reply sqs_workers/async_task.py:36]: _Rubber ducking this, a bit stream of conscious I'm afraid:_

In our readme we're clear that for non batched read messages kwargs should be used, as they are matched to the args in the call handler:

```python
@queue.processor("send_email")
def send_email(to, subject, body):
    print(f"Sending email {subject} to {to}")

send_email.delay(to="user@example.com", subject="Hello world", body="hello world")
```

Playing with `bind_arguments` this works, so refactored to this:

```python
kwargs = bind_arguments(self.processor, [[kwargs]], {})
return self.processor(**kwargs)
```

However in this instance we are still throwing away the args.

Because for batching we can't infer the parameter name from the underlying call handler, we wouldn't be able to map any args into kwargs based on position (which is what I believe `bind_arguments` is doing currently?)

So I've done as suggested and raising an error if `args` isn't empty for a batch task.

```python
if self.queue.batching_policy.batching_enabled:
    if len(args) > 0:
        raise TypeError("Must use keyword arguments only for batch read queues")
    kwargs = bind_arguments(self.processor, [[kwargs]], {})
    return self.processor(**kwargs)
else:
    kwargs = bind_arguments(self.processor, args, kwargs)
    return self.processor(**kwargs)
```

File: sqs_workers/async_task.py (hunk 8 above)
Comment (lukemerrett) [reply sqs_workers/async_task.py:62]: Done as outlined in the earlier comment

**Comment by lukemerrett** [comment]
Ready for re-review, with the outstanding logging comments to discuss @amleczko @Schnouki

**Review by goncalossilva — APPROVED** [review]
Same! This is @Schnouki's baby at this point. 😄

File: sqs_workers/processors.py (hunk 1 above)
Comment (Schnouki) [reply sqs_workers/processors.py]: Sure! That's why we have `extra` :)

**Review by Schnouki — APPROVED** [review]
Looks great, thanks a lot @lukemerrett! 🚀

**Comment by Schnouki** [comment]
> This is @Schnouki's baby at this point. :smile: 

More than happy to start co-parenting with @lukemerrett 😁

**Comment by lukemerrett** [comment]
🙌 I'm game 🙂

Thank you Thomas, will ping you about the release process
